libraries/custom_appbase/include/eosio/chain/application.hpp

```diff
@@ -81,13 +91,20 @@ class two_queue_executor {
       return more;
    }
 
-   bool execute_highest_read_only() {
+   bool execute_highest_read() {
       // execute for at least minimum runtime
       const auto end = std::chrono::high_resolution_clock::now() + std::chrono::milliseconds(minimum_runtime_ms);
 
       bool more = false;
       while (true) {
-         more = read_only_queue_.execute_highest_locked( true );
+         std::optional<bool> exec_read_only_queue = read_only_queue_.compare_queues_locked(read_exclusive_queue_);
+         if (!exec_read_only_queue) {
+            more = read_only_queue_.execute_highest_locked( true );
```

Comment:
I don't think I follow. `execute_highest_locked` returns `false` if queue is empty or if processing should stop.

Reply:
When `read_only_queue_.compare_queues_locked()` returns `{}`, it means that both queues are empty. So why call `read_only_queue_.execute_highest_locked( true );`, instead of just setting `more = false;`

Reply:
Ah. The `true` means block, so it allows it to block waiting for work instead of exiting loop.

Reply:
Hum, I think we should always pass `should_block` = true even in the two calls below. Otherwise thread1 can check and see the queue not empty, be interrupted by thread2 which dequeues the task, and when thread1 is resumed, it will call `execute_highest_locked` without `should_wait` , find an empty queue, return false and exit.

Reply:
Oh, there is a problem, because threads can be waiting on a queue while we add tasks to the other queue.

Reply:
Yes, there is an issue here. `more` should be true unless the blocking call returns `false`. I'm writing some tests for this now. Hopefully I can write a test that fails the current code.

Reply:
And seems like we need to pull the condition variable out of `exec_pri_queue` and put it in `three_queue_executor`.

Reply:
> And seems like we need to pull the condition variable out of `exec_pri_queue` and put it in `three_queue_executor`

I think I would modify `appbase::exec_pri_queue` to manage two `handlers` priority queue.... the logic could stay essentially the same.

Reply:
> I'm writing some tests

One issue, If you enter the read window with a single readonly trx enqueued, one thread will take it and all the other threads will wait on `read_only_queue_`, causing newly enqueued readonly trx to be processed in single thread mode.

Reply:
The new impl addresses this.